File: java/src/test/java/ai/dikestra/shield/V4VectorsTest.java
package ai.dikestra.shield;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Conformance: reproduce the Rust-generated v4 vectors byte-for-byte.
 *
 * Dependency-free: parses the small, well-formed tests/v4_test_vectors.json by
 * scanning each object inside the deterministic arrays and extracting fields with
 * regular expressions. Proves the Java binding derives identical master + AEAD
 * keys, reproduces every ciphertext byte-for-byte, and decrypts each vector.
 */
public class V4VectorsTest {

    private static Path vectorsPath() {
        Path dir = Paths.get("").toAbsolutePath();
        for (int i = 0; i < 8 && dir != null; i++) {
            Path candidate = dir.resolve("tests").resolve("v4_test_vectors.json");
            if (Files.exists(candidate)) return candidate;
            dir = dir.getParent();
        }
        throw new RuntimeException("v4_test_vectors.json not found");
    }

    private static String str(String obj, String field) {
        Matcher m = Pattern.compile("\"" + field + "\"\\s*:\\s*\"([^\"]*)\"").matcher(obj);
        return m.find() ? m.group(1) : null;
    }

    private static long num(String obj, String field) {
        Matcher m = Pattern.compile("\"" + field + "\"\\s*:\\s*(\\d+)").matcher(obj);
        return m.find() ? Long.parseLong(m.group(1)) : 0L;
    }

    /** Extract each brace-balanced object inside the named JSON array. */
    private static List<String> objectsInArray(String json, String arrayName) {
        List<String> objects = new ArrayList<>();
        int key = json.indexOf("\"" + arrayName + "\"");
        if (key < 0) return objects;
        int start = json.indexOf('[', key);
        int depth = 0, objStart = -1;
        for (int i = start; i < json.length(); i++) {
            char c = json.charAt(i);
            if (c == '[') { depth++; }
            else if (c == ']') { depth--; if (depth == 0) break; }
            else if (c == '{') { if (objStart < 0) objStart = i; }
            else if (c == '}') {
                if (objStart >= 0) { objects.add(json.substring(objStart, i + 1)); objStart = -1; }
            }
        }
        return objects;
    }

    private static List<String> allVectors() throws IOException {
        String json = new String(Files.readAllBytes(vectorsPath()));
        List<String> all = new ArrayList<>(objectsInArray(json, "deterministic_vectors"));
        all.addAll(objectsInArray(json, "deterministic_vectors_chacha"));
        return all;
    }

    private static byte[] hex(String s) {
        int len = s.length();
        byte[] out = new byte[len / 2];
        for (int i = 0; i < len; i += 2) {
            out[i / 2] = (byte) Integer.parseInt(s.substring(i, i + 2), 16);
        }
        return out;
    }

    private static String toHex(byte[] b) {
        StringBuilder sb = new StringBuilder(b.length * 2);
        for (byte x : b) sb.append(String.format("%02x", x));
        return sb.toString();
    }

    private static byte suiteByte(String obj) {
        return "0x02".equals(str(obj, "suite")) ? Shield.SUITE_CHACHA20_POLY1305 : Shield.SUITE_AES_GCM;
    }

    private static byte[] masterFor(String obj) {
        if ("password".equals(str(obj, "mode"))) {
            Shield s = new Shield(str(obj, "password"), str(obj, "service"),
                    hex(str(obj, "salt_hex")), (int) num(obj, "iterations"), null);
            return s.getKey();
        }
        return hex(str(obj, "key_hex"));
    }

    @Test
    void testKdfVectors() throws IOException {
        List<String> vectors = allVectors();
        assertTrue(vectors.size() >= 6, "expected vectors loaded");
        for (String v : vectors) {
            byte[] master = masterFor(v);
            assertEquals(str(v, "master_key_hex"), toHex(master), "master drift " + str(v, "name"));
            assertEquals(str(v, "aead_key_hex"), toHex(Shield.deriveAeadKey(master)),
                    "aead drift " + str(v, "name"));
        }
    }

    @Test
    void testReproduceBytes() throws IOException {
        for (String v : allVectors()) {
            byte[] aeadKey = Shield.deriveAeadKey(masterFor(v));
            byte[] salt = "password".equals(str(v, "mode")) ? hex(str(v, "salt_hex")) : null;
            byte[] out = Shield.sealDeterministic(aeadKey, suiteByte(v), salt,
                    hex(str(v, "nonce_hex")), num(v, "timestamp_ms"), (int) num(v, "pad_len"),
                    hex(str(v, "padding_hex")), hex(str(v, "plaintext_hex")));
            assertEquals(str(v, "expected_output_hex"), toHex(out), "BYTE DRIFT " + str(v, "name"));
        }
    }

    @Test
    void testDecryptVectors() throws IOException {
        for (String v : allVectors()) {
            byte[] aeadKey = Shield.deriveAeadKey(masterFor(v));
            byte[] encrypted = hex(str(v, "expected_output_hex"));
            int aadLen = "password".equals(str(v, "mode")) ? 2 + Shield.SALT_SIZE : 2;
            byte[] opened = Shield.openCiphertext(aeadKey, suiteByte(v), encrypted, aadLen, null);
            assertEquals(str(v, "plaintext_hex"), toHex(opened), "decrypt failed " + str(v, "name"));
        }
    }
}

File: android/shield/src/test/java/ai/dikestra/shield/V4VectorsTest.kt
package ai.dikestra.shield

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.File

/**
 * Conformance: reproduce the Rust-generated v4 vectors byte-for-byte.
 *
 * Dependency-free: scans each object inside the deterministic arrays of
 * tests/v4_test_vectors.json and extracts fields with regular expressions.
 * Runs on the local JVM (testDebugUnitTest), so the JCE AEADs are available.
 */
class V4VectorsTest {

    private fun vectorsFile(): File {
        var dir: File? = File("").absoluteFile
        repeat(8) {
            if (dir != null) {
                val candidate = File(dir, "tests/v4_test_vectors.json")
                if (candidate.exists()) return candidate
                dir = dir!!.parentFile
            }
        }
        throw RuntimeException("v4_test_vectors.json not found")
    }

    private fun str(obj: String, field: String): String? =
        Regex("\"$field\"\\s*:\\s*\"([^\"]*)\"").find(obj)?.groupValues?.get(1)

    private fun num(obj: String, field: String): Long =
        Regex("\"$field\"\\s*:\\s*(\\d+)").find(obj)?.groupValues?.get(1)?.toLong() ?: 0L

    private fun objectsInArray(json: String, arrayName: String): List<String> {
        val objects = mutableListOf<String>()
        val key = json.indexOf("\"$arrayName\"")
        if (key < 0) return objects
        val start = json.indexOf('[', key)
        var depth = 0
        var objStart = -1
        var i = start
        while (i < json.length) {
            when (json[i]) {
                '[' -> depth++
                ']' -> { depth--; if (depth == 0) break }
                '{' -> if (objStart < 0) objStart = i
                '}' -> if (objStart >= 0) { objects.add(json.substring(objStart, i + 1)); objStart = -1 }
            }
            i++
        }
        return objects
    }

    private fun allVectors(): List<String> {
        val json = vectorsFile().readText()
        return objectsInArray(json, "deterministic_vectors") +
            objectsInArray(json, "deterministic_vectors_chacha")
    }

    private fun hex(s: String): ByteArray =
        ByteArray(s.length / 2) { s.substring(it * 2, it * 2 + 2).toInt(16).toByte() }

    private fun toHex(b: ByteArray): String = b.joinToString("") { "%02x".format(it) }

    private fun suiteByte(obj: String): Byte =
        if (str(obj, "suite") == "0x02") Shield.SUITE_CHACHA20_POLY1305 else Shield.SUITE_AES_GCM

    private fun masterFor(obj: String): ByteArray {
        return if (str(obj, "mode") == "password") {
            val salt = hex(str(obj, "salt_hex")!!)
            val pbkdf2Salt = salt + str(obj, "service")!!.toByteArray(Charsets.UTF_8)
            ShieldUtils.pbkdf2(str(obj, "password")!!, pbkdf2Salt, num(obj, "iterations").toInt(), 32)
        } else {
            hex(str(obj, "key_hex")!!)
        }
    }

    @Test
    fun testKdfVectors() {
        val vectors = allVectors()
        assertTrue("expected vectors loaded", vectors.size >= 6)
        for (v in vectors) {
            val master = masterFor(v)
            assertEquals("master drift ${str(v, "name")}", str(v, "master_key_hex"), toHex(master))
            assertEquals("aead drift ${str(v, "name")}", str(v, "aead_key_hex"), toHex(Shield.deriveAeadKey(master)))
        }
    }

    @Test
    fun testReproduceBytes() {
        for (v in allVectors()) {
            val aeadKey = Shield.deriveAeadKey(masterFor(v))
            val salt = if (str(v, "mode") == "password") hex(str(v, "salt_hex")!!) else null
            val out = Shield.sealDeterministic(
                aeadKey, suiteByte(v), salt, hex(str(v, "nonce_hex")!!),
                num(v, "timestamp_ms"), num(v, "pad_len").toInt(),
                hex(str(v, "padding_hex")!!), hex(str(v, "plaintext_hex")!!)
            )
            assertEquals("BYTE DRIFT ${str(v, "name")}", str(v, "expected_output_hex"), toHex(out))
        }
    }

    @Test
    fun testDecryptVectors() {
        for (v in allVectors()) {
            val aeadKey = Shield.deriveAeadKey(masterFor(v))
            val encrypted = hex(str(v, "expected_output_hex")!!)
            val aadLen = if (str(v, "mode") == "password") 2 + Shield.SALT_SIZE else 2
            val opened = Shield.openCiphertext(aeadKey, suiteByte(v), encrypted, aadLen, null)
            assertEquals("decrypt failed ${str(v, "name")}", str(v, "plaintext_hex"), toHex(opened))
        }
    }
}
